Replace react-vis with recharts

navaneeth-dev opened the pull request:
## Which problem is this PR solving?
- Resolves #1597

## Description of the changes
Still work in progress, copying all the changes from previous PR #2242 and implementing the needed tests.

## How was this change tested?
- `npm run start`
- `npm run build`
- `npm run test`

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `npm run lint` and `npm run test`

navaneeth-dev commented:
@yurishkuro My first step is replacing on the search page and get the tests working, then move on to other pages.

Changes made that are different from the original PR:
1. Fixes non-uniform time: Max 4 ticks, interval evenly calculated from start and end in `generateTicks()`
2. Made Y Axis label "Trace duration"

![image](https://github.com/user-attachments/assets/56d82a2b-756d-48e1-b92d-a3f6f502c8e6)
---
![image](https://github.com/user-attachments/assets/2e51d306-2400-4351-9a1b-37478f8f2186)

Which should we use? I tried to replicate many things from the current search page like colors, font size.
The current version has behavior like the where the cells go outside the graph.

TODO: Working on making cell size similar to the current behavior.

Thanks in advance.

yurishkuro commented:
> Grafana does intervals according to scrape_interval in Prometheus 

but what if I select a larger time range? Doing ticks at scrape intervals would cause labels to overlap.

yurishkuro commented:
please make sure you work off up to date main branch, you already have merge conflicts

yurishkuro commented:
> Do you want Grafana like behavior?

I don't care about it being specifically Grafana-like, just that the ticks need to adjust depending on the time range of the query, but also the screen size. E.g. in your screenshot you can easily go from 4 to 8 ticks, which will increase readability, but if I start reducing the screen width the 8 ticks might start overlapping, so what is the solution to that?

navaneeth-dev commented:
I tried the `interval={equidistantPreserveStart}` option, but it is NOT equidistant in the middle.

![image](https://github.com/user-attachments/assets/402a6466-e7d9-4b2f-96db-5ee02e95580e)

There is an issue open but no timeline: https://github.com/recharts/recharts/issues/3305

I think for now we have to manually calculate it.

yurishkuro commented:
@navaneeth-dev lmk when ready for review

navaneeth-dev commented:
Will do, working on migrating tests from emyze to react-testing-library.

Have you thought about adding .editorconfig?

yurishkuro commented:
> working on migrating tests from emyze to react-testing-library.

is there a lot of tests? Perhaps we can merge them in a separate PR before merging this one.

> Have you thought about adding .editorconfig?

I don't know what that is, sounds like specific to an IDE that you are using.

navaneeth-dev commented:
No just started work on `ScatterPlot.test.js`, was busy yesterday. So do you want me to finish ScatterPlot component fully with tests and let know know for this PR?

---

"Editorconfig helps maintain consistent coding styles for multiple developers working on the same project across various editors and IDEs": 

So formatter (prettier) usually comes after saving, but EditorConfig sets up the editor's: space, indent size correctly etc...
I highly recommend it and many CNCF projects use it. Currently, I have it git excluded with Neovim.
https://editorconfig.org/

yurishkuro commented:
my point was that changing tests to `react-testing-library` is a task on its own, it's good to do it _before_ making changes to the code, such that we can then actually just the later code changes based on whether the tests are passing.

navaneeth-dev commented:
Understood, will migrate the remaining files and let you know:

./packages/jaeger-ui/src/components/Monitor/ServicesView/serviceGraph.tsx
./packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/opsGraph.tsx

navaneeth-dev commented:
What is the purpose of a Line/LineSeries here? In my demo instance, I can't see any lines.
https://github.com/MarkPollmann/jaeger-ui/blob/e8c58019494cc755bc0f6efad86af5eb00f9646f/packages/jaeger-ui/src/components/Monitor/ServicesView/serviceGraph.tsx#L93

navaneeth-dev commented:
@yurishkuro 

Service Graph latency, error, request rate graphs finished. Is the tooltip fine?

https://github.com/user-attachments/assets/aff7b54a-de75-473d-9048-9e2199e18b99

Bottom graphs seem to have some bug, will be fixing that.
![image](https://github.com/user-attachments/assets/a5a6af01-2b06-4092-87e7-aff006cdde6a)

navaneeth-dev commented:
Fixed opsGraph, this PR is now **ready**.

I have two questions:
1. Error rate data point's y value IS 0, so I set minimum value to 0.1 for every opsGraph so it displays a red line.
2. Do we need sharp steps like below OR like my previous video which is linear? When the service is down, there is no data so there is a gap.

![image](https://github.com/user-attachments/assets/740844ed-18b4-4193-b192-064746a67f9b)

EDIT: I fixed the order of time to behave like current main one.

yurishkuro commented on packages/jaeger-ui/src/components/Monitor/ServicesView/index.tsx:
```diff
@@ -332,10 +334,8 @@ export class MonitorATMServicesViewImpl extends React.PureComponent<TProps, Stat
-                    }h&maxDuration&minDuration&service=${this.getSelectedService()}&start=${
-                      Date.now() - selectedTimeFrame
+                    `/search?end=${Date.now()}000&limit=20&lookback=${selectedTimeFrame / (3600 * 1000)
+                    }h&maxDuration&minDuration&service=${this.getSelectedService()}&start=${Date.now() - selectedTimeFrame
```
changes to this file appear to be non-functional, please revert

navaneeth-dev replied:
I think my formatter made the 4 lines into 2, will revert

yurishkuro commented on packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/opsGraph.tsx:
```diff
@@ -51,34 +46,28 @@ export class OperationsGraph extends React.PureComponent<TProps> {
 
     if (yDomain) {
-      dynProps.yDomain = yDomain;
+      dynProps.domain = yDomain;
```
please avoid unnecessary changes, I want to focus my review on the main changes, not be distracted by these ^^^

navaneeth-dev replied:
This is a necessary change, `<YAxis hide={true} {...dynProps} />` YAxis takes domain prop not `yDomain` like `react-vis`.

How would you want me to proceed?

yurishkuro replied:
do you need this const and if at all? Could it not be just inlined where needed?

yurishkuro commented on packages/jaeger-ui/src/components/SearchTracePage/SearchResults/ScatterPlot.jsx:
```diff
@@ -15,28 +15,34 @@
   const [containerWidth, setContainerWidth] = useState(0);
 
-  const [overValue, setValueOver] = useState(null);
+  const overValue = useRef();
```
why is this change needed?

navaneeth-dev replied:
It's not needed, it was pasted code from previous PR. Fixed in latest commit.

yurishkuro reviewed (commented):
please resolve conflicts and update branch up to date

navaneeth-dev reviewed (commented):
Fixed all merge conflicts and updated to main.

yurishkuro commented on packages/jaeger-ui/package.json:
```diff
@@ -82,6 +82,7 @@
     "react-vis-force": "^0.3.1",
```
is react-vis still needed after this change?

navaneeth-dev replied:
No, I will remove it

yurishkuro reviewed (commented):
I ran monitor example. The Search screen did not show scatterplot at all

<img width="881" alt="image" src="https://github.com/user-attachments/assets/44937cba-d3ea-428c-b89b-baaacaebe7a0" />

The Monitor page showed charts, but their time axis doesn't make sense, when I switched to 24hrs view:

<img width="1145" alt="image" src="https://github.com/user-attachments/assets/8e8926b4-a75c-4f1a-aec4-aa01da5c8783" />

navaneeth-dev commented:
react-vis-force is different from react-vis. Once we migrate the tests we can remove react-vis.

```bash
rg react-vis                                     
package.json
49:    "prettier": "prettier --write '{.,scripts}/*.{js,jsx,json,md,ts,tsx,mts}' 'packages/*/{src,demo/src}/**/!(layout.worker.bundled|react-vis).{css,js,jsx,json,md,ts,tsx}' 'packages/*/*.{css,js,jsx,json,md,ts,tsx,mts}'",
50:    "prettier-lint": "prettier --list-different '{.,scripts}/*.{js,jsx,json,md,ts,tsx,mts}' 'packages/*/{src,demo/src}/**/!(layout.worker.bundled|react-vis).{css,js,jsx,json,md,ts,tsx}' 'packages/*/*.{css,js,jsx,json,md,ts,tsx,mts}' || (echo '*** PLEASE RUN npm run prettier AND RESUBMIT ***' && false)",

packages/jaeger-ui/package.json
81:    "react-vis": "1.11.12",
82:    "react-vis-force": "^0.3.1",

packages/jaeger-ui/typings/custom.d.ts
43:declare module 'react-vis-force';

package-lock.json
16602:    "node_modules/react-vis": {
16604:      "resolved": "https://registry.npmjs.org/react-vis/-/react-vis-1.11.12.tgz",
16635:    "node_modules/react-vis-force": {
16637:      "resolved": "https://registry.npmjs.org/react-vis-force/-/react-vis-force-0.3.1.tgz",
16651:    "node_modules/react-vis/node_modules/d3-color": {
16657:    "node_modules/react-vis/node_modules/d3-interpolate": {
16666:    "node_modules/react-vis/node_modules/deep-equal": {
20264:        "react-vis": "1.11.12",
20265:        "react-vis-force": "^0.3.1",

packages/jaeger-ui/src/components/DependencyGraph/ForceGraphArrowLink.tsx
22:import { ForceGraphLink } from 'react-vis-force';

packages/jaeger-ui/src/components/DependencyGraph/DependencyForceGraph.test.js
17:import { InteractiveForceGraph, ForceGraphNode } from 'react-vis-force';

packages/jaeger-ui/src/components/DependencyGraph/DependencyForceGraph.jsx
16:import { InteractiveForceGraph, ForceGraphNode } from 'react-vis-force';

packages/jaeger-ui/src/components/SearchTracePage/SearchResults/ScatterPlot.test.js
17:import { XAxis, XYPlot, YAxis, MarkSeries, Hint } from 'react-vis';
```

navaneeth-dev commented:
I think there is a bug with the size of ScatterPlot circles will test and let you know. I ran monitor docker compose example too but it shows for me :thinking: .

As for Monitor I used default recharts xaxis so interval isn't uniform. I will try to make it like Grafana and let you know.

yurishkuro commented:
> react-vis-force is different from react-vis.

I know, I was referring to react-vis only.

yurishkuro commented:
Addressed in #2679, but thanks for trying anyway.
